src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java

```diff
@@ -194,6 +195,18 @@ public DeploymentResult call() throws InterruptedException {
         Optional<Set<String>> groupsForDeviceOpt;
         try {
             groupsForDeviceOpt = thingGroupHelper.listThingGroupsForDevice(retryCount);
+        } catch(GreengrassV2DataException e) {
+            if (e.statusCode() == 403) {
+                // Getting group hierarchy requires permission to call the ListThingGroupsForCoreDevice API which
+                // may not be configured on existing IoT Thing policy in use for current device, log a warning in
+                // that case and move on.
+                logger.atWarn().setCause(e).log("Failed to get thing group hierarchy. Deployment will proceed. "
+                        + "To automatically clean up unused components, please add "
+                        + "greengrass:ListThingGroupsForCoreDevice permission to your IoT Thing policy.");
+                groupsForDeviceOpt = getPersistedMembershipInfo();
+            } else {
+                throw new DeploymentTaskFailureException("Error fetching thing group information", e);
```

Comment:
As an aside, what gets reported as the error in the deployment status? Do we follow the whole exception chain and generate a message or what are we doing? If we're only taking the top exception of "Error fetch thing group information" I think we could improve that experience.

Reply:
It looks like we only include the top level message due to the way the task failure exception is wrapped around the original error, so yeah we should definitely improve it. Will take that up separately since this is actually avoiding a deployment failure right now